### PR TITLE
feat(config): add docker cli config

### PR DESCRIPTION
**Pull request by nabeelsaabna**
## Changes

Add new `dockerCliOptions` global config

## Context

https://github.com/renovatebot/renovate/issues/22573

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/options/index.ts [hunk 1]
```diff
@@ -343,6 +343,12 @@ const options: RenovateOptions[] = [
     globalOnly: true,
     default: 'renovate_',
   },
+  {
+    name: 'dockerCliOptions',
+    description: 'Change this value to add docker run command options to Renovate Docker sidecar container.',
```
Comment (rarkins) [inline lib/config/options/index.ts:348]:
```suggestion
    description: 'Docker run command options for use when `binarySource=docker`.',
```

File: docs/usage/self-hosted-configuration.md [hunk 2]
```diff
@@ -328,6 +328,16 @@ For example, if you set `dockerChildPrefix=myprefix_` then the final container c
 !!! note
     Dangling containers are only removed when Renovate runs again with the same prefix.
 
+## dockerCliOptions
+
+Adds command-line docker cli options to Renovate sidecar Docker containers.
```
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md:333]:
```suggestion
Adds command-line docker CLI options to Renovate sidecar Docker containers.
```

File: docs/usage/self-hosted-configuration.md [hunk 3]
```diff
@@ -328,6 +328,16 @@ For example, if you set `dockerChildPrefix=myprefix_` then the final container c
...
+
+for example `--memory=1g --cpus="1.5"`
```
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md:335]:
```suggestion
For example, `--memory=1g --cpus="1.5"`.
```

File: docs/usage/self-hosted-configuration.md [hunk 4]
```diff
@@ -328,6 +328,16 @@ For example, if you set `dockerChildPrefix=myprefix_` then the final container c
...
+
+Read the [Docker runtime resource constraints](https://docs.docker.com/config/containers/resource_constraints/) for more information.
+
+Note: there are other config options available for docker side-car container that adds cli options as well for running and also cleaning containers from previous runs
+
```
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md:340]:
```suggestion
```

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by viceice — COMMENTED** [review]
code LGTM

File: docs/usage/self-hosted-configuration.md [hunk 5]
```diff
@@ -328,6 +328,14 @@ For example, if you set `dockerChildPrefix=myprefix_` then the final container c
 !!! note
     Dangling containers are only removed when Renovate runs again with the same prefix.
 
+## dockerCliOptions
+
+Adds command-line docker CLI options to Renovate sidecar Docker containers.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:333]:
```suggestion
You can use `dockerCliOptions` to pass Docker CLI options to Renovate's sidecar Docker containers.
```

File: docs/usage/self-hosted-configuration.md [hunk 6]
```diff
@@ -328,6 +328,14 @@ For example, if you set `dockerChildPrefix=myprefix_` then the final container c
...
+
+For example, `--memory=1g --cpus="1.5"`.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:335]: I'd like a real config option example here, something similar to:

```json
{
  "dockerCliOptions": ["this-is-a-bad-config-do-not-copy"]
}
```

File: docs/usage/self-hosted-configuration.md [hunk 7]
```diff
@@ -328,6 +328,14 @@ For example, if you set `dockerChildPrefix=myprefix_` then the final container c
...
+
+Read the [Docker runtime resource constraints](https://docs.docker.com/config/containers/resource_constraints/) for more information.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:337]: It feels like we're missing an explanation on our side here. How about a structure like this:

```markdown
One line summary of config option.

Explanation about _why_ and _when_ a user may need this config option.
Expand on the `--memory=1g --cpus="1,5"` stuff.

Real config example in proper JSON.

To learn more about ... read the Docker runtime resource constrains.
```

File: lib/config/options/index.ts [hunk 8]
```diff
@@ -343,6 +343,12 @@ const options: RenovateOptions[] = [
     globalOnly: true,
     default: 'renovate_',
   },
+  {
+    name: 'dockerCliOptions',
+    description: 'Docker run command options for use when `binarySource=docker`.',
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:348]: This description needs to be improved as well. It feels a bit too short for me.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/self-hosted-configuration.md (hunk 6 above)
Comment (rarkins) [reply docs/usage/self-hosted-configuration.md:335]: It's not completely bad or good

Comment (viceice) [reply docs/usage/self-hosted-configuration.md:335]: 🤔  think those are already valid options.

File: docs/usage/self-hosted-configuration.md (hunk 7 above)
Comment (nabeelsaabna) [reply docs/usage/self-hosted-configuration.md]: updated

**Comment by nabeelsaabna** [comment]
1. Tested e2e successfully with `dockerCliOption: --memory=1g` updating `packge-lock.json` and `go.sum` 
2. Test passed with the expected error with  `dockerCliOption: --memory=7m` and error:

> 
> ### ⚠ Artifact update problem
> 
> Renovate failed to update an artifact related to this branch. You probably do not want to merge this PR as-is.
> 
> ♻ Renovate will retry this branch, including artifacts, only when one of the following happens:
> 
>  - any of the package files in this branch needs updating, or 
>  - the branch becomes conflicted, or
>  - you click the rebase/retry checkbox if found above, or
>  - you rename this PR's title to start with "rebase!" to trigger it manually
> 
> The artifact failure details are included below:
> 
> ##### File name: go.sum
> 
> ```
> Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child --memory=7m -v "/tmp/renovate/repos/github/nabeelsaabnaTests/test-new-job":"/tmp/renovate/repos/github/nabeelsaabnaTests/test-new-job" -v "/tmp/renovate/cache":"/tmp/renovate/cache" -e GOPATH -e GOFLAGS -e CGO_ENABLED -e GIT_CONFIG_KEY_0 -e GIT_CONFIG_VALUE_0 -e GIT_CONFIG_KEY_1 -e GIT_CONFIG_VALUE_1 -e GIT_CONFIG_KEY_2 -e GIT_CONFIG_VALUE_2 -e GIT_CONFIG_COUNT -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w "/tmp/renovate/repos/github/nabeelsaabnaTests/test-new-job" docker.io/containerbase/sidecar bash -l -c "install-tool golang 1.20.5 && go get -d -t ./..."
> bash: line 1:    10 Killed                  install-tool golang 1.20.5
> 
> ```
> 
>

File: lib/config/options/index.ts (hunk 8 above)
Comment (nabeelsaabna) [reply lib/config/options/index.ts]: not sure what to add🤔

**Review by rarkins — DISMISSED** [review]

File: docs/usage/self-hosted-configuration.md [hunk 9]
```diff
@@ -333,6 +333,14 @@ For example, if you set `dockerChildPrefix=myprefix_` then the final container c
 !!! note
     Dangling containers are only removed when Renovate runs again with the same prefix.
 
+## dockerCliOptions
+
+You can use `dockerCliOptions` to pass Docker CLI options to Renovate's sidecar Docker containers.
+
+For example, `{"dockerCliOptions": "--memory=4g"}` will add a cli flag to the docker run command that will cap the maximum amount of memory Renovate's sidecar Docker container can use to 4 gigabytes.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:340]:
```suggestion
For example, `{"dockerCliOptions": "--memory=4g"}` will add a CLI flag to the `docker run` command that limits the amount of memory Renovate's sidecar Docker container can use to 4 gigabytes.
```

File: docs/usage/self-hosted-configuration.md [hunk 10]
```diff
@@ -333,6 +333,14 @@ For example, if you set `dockerChildPrefix=myprefix_` then the final container c
...
+
+Click here to learn more about [Docker runtime resource constraints](https://docs.docker.com/config/containers/resource_constraints/).
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:342]:
```suggestion
Read the [Docker Docs, configure runtime resource contraints](https://docs.docker.com/config/containers/resource_constraints/) to learn more.
```

File: lib/config/options/index.ts [hunk 11]
```diff
@@ -343,6 +343,13 @@ const options: RenovateOptions[] = [
     globalOnly: true,
     default: 'renovate_',
   },
+  {
+    name: 'dockerCliOptions',
+    description:
+      'Docker run command options for use when `binarySource=docker`.',
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:349]:
```suggestion
      'Pass CLI flags to `docker run` command when `binarySource=docker`.',
```

I think we should put `docker run` in a codeblock style, so it's clear it's about the `docker run` command.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by viceice — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**